### ISSUE TITLE
Silent data loss

### ISSUE THREAD

**Issue by zaoxing** [ProjectDMX/DMI#100]
# Silent activation loss: one failed ClickHouse insert kills the sink for the rest of the run

**Type:** bug (data loss / silent corruption)
**Component:** native host sink (`DMXHostEngine`), ring p2p thread, `dmi.storage.internals`
**Affects:** both HF and vLLM paths — the defect is in the shared native sink, not in either integration
**Found by:** static reading of `main` @ `3f95d87`. Not yet reproduced on hardware; a suggested repro is in §4.

## 1. Summary

A single transient ClickHouse insert error permanently disables the capture sink for the remainder of the
process. The failure is announced by exactly one `stderr` warning, is never surfaced to Python, and is not
detected by the default read path. `generate()` returns correct text, the process exits 0, and the
activation table is silently truncated from the moment of the error onward.

The isolation itself is correct and intended — the sink must never block or crash inference. The defect is
narrower: **isolation was implemented as suppression rather than as accounting.** Not blocking the GPU and
not telling anyone are separable concerns; DMI currently does both.

## 2. Root cause chain

1. **A failed insert aborts the whole engine, not just the batch.**
   `client.Insert` throws at `native/csrc/clickhouse_client.cpp:514`. The stage worker catches it, records a
   `ThreadFailure`, and calls `maybe_abort_on_failure_()` (`native/csrc/pipelined_engine.hpp:931`). The
   default policy is `ThreadExceptionPolicy::STOP_ENGINE` (`pipelined_engine.hpp:124`), which calls
   `request_abort()` (`pipelined_engine.hpp:744`). `DMXHostEngine` constructs with a default
   `EngineConfig{}` (`native/csrc/dmx_host_engine.h:15`), so it inherits that policy. The in-flight batch is
   never retried and the sink is dead for the rest of the process.

2. **Every subsequent submit throws.**
   `submit_items` checks the abort flag up front and raises `"Engine is stopping/aborted"`
   (`pipelined_engine.hpp:586`).

3. **Those throws are suppressed after the first.**
   The p2p thread wraps each submit and routes failures into a `std::once_flag` logger
   (`native/csrc/ring/p2p_thread.cpp:310`, `:20`), which prints one line for the entire process lifetime:
   `"[DMI][P2P] WARN: failed to submit tensor slice to host engine; suppressing further submit errors."`
   The wording reads as a per-slice hiccup, not as "the sink is gone."

4. **Python never asks.**
   `raise_if_failed()` and `failures()` are bound (`native/csrc/bindings.cpp:215`) but consulted nowhere in
   `src/` — the only repo hits are `pass` stubs in benchmark mocks (`benchmarks/bench_hf_transport.py:84`,
   `benchmarks/profile_decode.py:72`). The one natural detection point, teardown, swallows it
   (`src/dmi/engine.py:306`):

   ```python
   if self._host_engine is not None:
       try:
           self._host_engine.close_input()
           self._host_engine.stop()
       except Exception:
           pass
   ```

5. **The default read path does not validate completeness.**
   `InternalRequirement.match_token_ranges` defaults to `False` (`src/dmi/storage/internals.py:201`),
   `_validate_token_ranges` early-returns when `requirement is None`, and `get_internal(model_id)` attaches
   no requirements at all. The documented entry point therefore performs zero completeness checking. Even
   when a caller opts in, only one layer is inspected — `check_layer = layers[-1]`
   (`internals.py:449`), a fast path aimed at the "writer still pending" race, which does not align with a
   dropped batch spanning arbitrary layers.

6. **Sink telemetry is compiled out.**
   `PipelinedEngine` tracks `dropped`, `full_errors`, `retries` and per-stage timings
   (`pipelined_engine.hpp:249`, `:276`) behind `EnableEngineProfiling`, which `DMXHostEngine` instantiates as
   `false` (`dmx_host_engine.h:11`). `profiling()` returns `nullopt`, and nothing is bound to Python
   regardless. A run cannot report how much it lost.

## 3. Impact

A transient ClickHouse error at token 50 of a 10,000-token run discards every activation after token 50.
Observable symptoms: correct generated text, exit code 0, one `stderr` warning, and a truncated table that
the default read path reports as complete. Any downstream analysis — interpretability sweeps, distillation
datasets, steering experiments — consumes the truncated data as if it were whole.

This is worse than a crash: a crash is recoverable, a silently short dataset is not detectable after the
fact without an independent record of what should have been captured.

## 4. Suggested reproduction

Not yet run; requires a GPU host and a ClickHouse instance.

1. Start a capture run with persistence enabled (`dmx_db_host` set) and a long generation.
2. Mid-run, force one insert to fail — e.g. `SYSTEM SHUTDOWN` / restart the ClickHouse container, revoke
   `INSERT` on the target table, or `DETACH TABLE default.offload`.
3. Restore ClickHouse and let the run finish.
4. Observe: `generate()` completes normally, exit code 0, a single `[DMI][P2P] WARN` line.
5. `SELECT max(end_token_idx) FROM default.offload WHERE model_id=...` stops at the failure point; every
   later token is absent, and `get_internal(model_id)` raises nothing.

## 5. Why insert retry alone is not the fix

The table is plain `MergeTree` (`native/csrc/clickhouse_client.cpp:294`), so a retried insert can duplicate
rows on the sort key. Duplicates fail silently rather than loudly:

- `_actual_ranges_for_request` builds a **set** of `(start, end)` (`internals.py:426`), so duplicates
  collapse and validation passes.
- `_reassemble_per_layer` builds a **list** and `torch.cat`s it (`internals.py:66`), so a duplicated chunk
  lengthens the sequence axis.

Adding retry before idempotency would trade visible loss for invisible corruption. Ordering matters.

## 6. Proposed fix, in dependency order

1. **Count suppressed submits.** Keep the `once_flag` for the log line, but increment a counter on every
   suppressed submit in `p2p_thread.cpp`. Enable `EnableEngineProfiling` for `DMXHostEngine` and bind
   `StatsSnapshot` to Python so a run can report `dropped`/`full_errors`/`retries`.
2. **Surface sink death at a safe point.** Do not raise from the p2p thread. Check `failures()` in
   `close_input()`/`stop()` and let teardown warn or raise, instead of `except Exception: pass` at
   `engine.py:306`.
3. **Make inserts idempotent.** `ReplacingMergeTree` on the full sort key, or `insert_deduplication_token`;
   dedupe by key in the reassemblers.
4. **Then add bounded retry.** Set the insert stage's `exception_policy` to `CONTINUE` with retry, so a
   transient error costs one batch rather than the run.
5. **Flip the read-path default.** `match_token_ranges=True` by default for `get_internal()`; silent
   truncation is a worse outcome than a raised `IncompleteInternalError`. Add a strict mode that checks all
   layers rather than only the last.

Steps 1–2 are small, touch neither the schema nor the hot path, and are independently valuable: they convert
a silent failure into a loud one.

## 7. Non-goals

- Do not make the sink block or backpressure the GPU producer. The never-block property is the design
  thesis and should be preserved; this issue is about reporting, not about flow control.
